libmamba/src/solv-cpp/repo.cpp

```diff
@@ -83,34 +83,74 @@ namespace mamba::solv
 
     namespace
     {
-        struct FilePtr
+        class FilePtr
         {
-            std::FILE* ptr = nullptr;
-            std::string_view name = {};
+        public:
 
-            ~FilePtr() noexcept(false)
+            /**
+             * Open a file with C API.
+             *
+             * @param path must hae filesystem default encoding.
+             */
+            static auto open(std::filesystem::path path, const char* mode) -> FilePtr;
+
+            ~FilePtr() noexcept(false);
+
+            auto raw() noexcept -> std::FILE*;
+            auto raw() const noexcept -> const std::FILE*;
+
+        private:
+
+            FilePtr(std::FILE* ptr, std::string name);
+
+            std::FILE* m_ptr = nullptr;
+            std::string m_name = {};
+        };
+
+        FilePtr::FilePtr(std::FILE* ptr, std::string name)
+            : m_ptr{ ptr }
+            , m_name{ name }
+        {
+        }
+
+        auto FilePtr::open(std::filesystem::path path, const char* mode) -> FilePtr
+        {
+            std::string name = path;
+            std::FILE* ptr = std::fopen(name.c_str(), mode);
+            if (ptr == nullptr)
             {
-                const auto close_res = std::fclose(ptr);  // This flush too
-                if (close_res != 0)
-                {
-                    // TODO(C++20) fmt::format
-                    auto ss = std::stringstream();
-                    ss << "Unable to close file " << name;
-                    throw std::runtime_error(ss.str());
-                }
+                throw std::system_error(errno, std::generic_category());
             }
-        };
+            return { ptr, std::move(name) };
+        }
+
+        FilePtr::~FilePtr() noexcept(false)
```

Comment:
Allowing throwing destructor is definitely a recipe for failure and make some debugging harder.
Instead of `noexcept(false)`, I would recommend logging the error as error (if you want to provide a special error call funciton for this almost-separate-lib, do that) and just finish the destructor - no throw.

Reply:
I would prefer throwing than logging. Since this is only used as a local variable in a function, I would argue this is OKish

Reply:
😬

Reply:
@Klaim Otherwise, I would `if(m_ptr == nullptr) std::terminate()` in `~FilePtr` to force the user to call a throwing `FilePtr::close`, a la `std::tread`.